src/dmd/globals.d

```diff
@@ -529,17 +529,25 @@ alias d_uns64 = uint64_t;
 // file location
 struct Loc
 {
+nothrow:
     const(char)* filename; // either absolute or relative to cwd
-    uint linnum;
-    uint charnum;
+    private uint _linnum;
+    private uint _charnum;
 
     static immutable Loc initial;       /// use for default initialization of const ref Loc's
 
-nothrow:
+    @safe @nogc pure @property extern(C++)
+    {
+        const uint linnum() { return _linnum; }
+        const uint charnum() { return _charnum; }
+        void linnum(uint rhs) { _linnum = rhs; }
+        void charnum(uint rhs) { _charnum = rhs; }
+    }
```

Comment:
Sorry, but this looks like no improvement to me at all. And as long as the C++ versions aren't defined inline, LDC and GDC will require cross-language LTO to get rid of these calls.

Reply:
Yah, those are seldom called and never in core loops. I'd be glad to inline them if they ever become a problem.

The idea here is to encapsulate the data so as to look into alternative solutions (24/8 bits for line/column, doped vectors, global table, etc). Reducing footprint is a path worth pursuing. 

It's possible the final solution will not even need this diff.

Reply:
Until such a compression scheme is actually used, I consider this an unnecessary complication. Plus keeping the `filename` pointer as a public member will make any work on line/col in vain (struct guaranteed to be padded to 16 bytes for 64-bit code). - Iain had some ideas in the past IIRC, to compress a Loc into a single machine word.

Reply:
I've already tried this. If the size of `Loc` is reduced, the lexer gets faster. I tried lexing a file that is 140MB. It went from around 607 (or 647, don't remember which) milliseconds to 557. I measured by compiling with LDC.

I also tried with a global table. If I used a static preallocated table with enough entires for all locations, the speed was the same. If I used dynamic allocation, the lexer became slower than the original code.

BTW, we should try to avoid adding more global state, as much as possible. It makes it more difficult to use DMD as a library.

If the size of `Loc` can be reduced to 4 or 8 bytes it's small enough to be able to pass by value instead of by reference. Which might speed up things.

Reply:
> Iain had some ideas in the past IIRC, to compress a Loc into a single machine word.

In Clang the corresponding struct is 4 bytes. If I understand things correctly, it's more or less an index into an array. But if we start dynamically allocating this array the lexer will get slower (I haven't measured any other parts of the compiler). See my post above.